src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,713 @@
+{
+	evdev_device* mapped;
+	char name[64];
+	char id[32];
+	device_type type;
+	unsigned char num;
+	unsigned char buttons[OCTETS_FOR(KEY_MAX)];
+	unsigned char relaxes[OCTETS_FOR(REL_MAX)];
+	unsigned char absaxes[OCTETS_FOR(ABS_MAX)];
+};
+
+
+class evdev_device: public device_info
+{
+private:
+	// This is _way_ too conservative, but memory is cheap
+	const evdev_device_info& info;
+	int	fd;
+	uint8_t	keystate[KEY_MAX];
+	int32_t	relaxis[REL_MAX];
+	struct {
+		int32_t	normalized;
+		input_absinfo absinfo;
+	}	absaxis[ABS_MAX];
+
+public:
+	evdev_device(running_machine& machine, std::string&& name, std::string&& id, input_module& module,
+			input_device_class deviceclass, const evdev_device_info& deviceinfo):
+		device_info(machine, std::move(name), std::move(id), deviceclass, module),
+		info(deviceinfo)
+	{
+		fd = -1;
+		reset();
+	}
+
+	~evdev_device()
+	{
+		if(fd >= 0)
+			close(fd);
+	}
+
+	void start(int dirfd)
+	{
+		int numkeys = 0;
+		int numbuttons = 0;
+		int mapi = 0;
+		for(int i=1; i<KEY_MAX; i++)
+			if(info.buttons[i>>3] & (1<<(i&7))) {
+				while(key_map[mapi].evdev_id && key_map[mapi].evdev_id<i)
+					mapi++;
+				char name[32];
+				input_item_id id = ITEM_ID_INVALID;
+				name[0] = 0;
+				if(key_map[mapi].evdev_id == i) {
+					strncpy(name, key_map[mapi].name, 31);
+					name[31] = 0;
+					if((id = key_map[mapi].mame_id) != ITEM_ID_INVALID)
+						numkeys++;
+				} else {
+					if(i<BTN_MISC)
+						snprintf(name, 32, "KEYCODE %d", i);
+					else
+						snprintf(name, 32, "BUTTON %d", numbuttons+1);
+				}
+				if(id == ITEM_ID_INVALID && numbuttons<32)
+					id = static_cast<input_item_id>(ITEM_ID_BUTTON1 + numbuttons++);
+				if(id != ITEM_ID_INVALID) {
+					device()->add_item(name, id, generic_button_get_state<uint8_t>, keystate+i);
+				}
+			}
+
+		if(fd < 0) {
+			char fname[16];
+			snprintf(fname, 16, "event%d", info.num);
+			fd = openat(dirfd, fname, O_RDONLY|O_NONBLOCK);
+			if(fd < 0) {
+				osd_printf_verbose("evdev: unable to open %s: %s\n", fname, strerror(errno));
+			}
+		}
+
+		mapi = 0;
+		for(int i=0; i<REL_MAX; i++)
+			if(info.relaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || relaxis_map[mapi].evdev_id) && relaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(relaxis_map[mapi].evdev_id == i) {
+					device()->add_item(relaxis_map[mapi].name, relaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, relaxis+i);
+				}
+			}
+
+		mapi = 0;
+		for(int i=0; i<ABS_MAX; i++)
+			if(info.absaxes[i>>3] & (1<<(i&7))) {
+				while((mapi==0 || absaxis_map[mapi].evdev_id) && absaxis_map[mapi].evdev_id<i)
+					mapi++;
+				if(absaxis_map[mapi].evdev_id == i) {
+					if(fd >= 0)
+						ioctl(fd, EVIOCGABS(i), &absaxis[i].absinfo);
+					device()->add_item(absaxis_map[mapi].name, absaxis_map[mapi].mame_id,
+						generic_axis_get_state<int32_t>, &absaxis[i].normalized);
+				}
+			}
+	}
+
+	void poll(void) override
+	{
+		static input_event ie[256]; // try to consume as many as possible at once
```

Comment:
Having a function static like this is generally a bad sign.

Reply:
I wanted to avoid stack trashing on a function that is called really often; the actual lifetime of the buffer is exactly the life of the function call so making it local won't break things.

That said, my first obvious improvement is to move the read()s away from poll() entirely and have a thread that does a select()/read() asynchronously (and then only when there's an actual event) so it's kinda moot.

... which raises the question on whether this is a reasonable thing to do at all given your current architecture, and if it is what the "proper" way of doing it is.  I didn't find a good venue where I could ask questions like this to the dev team short of submitting WIP.  :-D

Reply:
This is now moot with the new worker thread